### PR TITLE
[SPARK-39522][INFRA]Uses Docker image cache over a custom image in pyspark job

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?
image

Change pyspark container from original static image to just-in-time build image from cache.

See also: https://docs.google.com/document/d/1_uiId-U1DODYyYZejAZeyz2OAjxcnA-xfwjynDF6vd0

This patch has 2 changes:
1. Add a `infra-image` job to build ci image from [cache images](https://github.com/apache/spark/pkgs/container/spark%2Fapache-spark-github-action-image-cache)
2. Use the ci image as pyspark job container image, only master branch enable the new workflow now.

### Why are the changes needed?
Help to speed up docker infra image build in each PR.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
CI passed

### CODE REVIEW

File: dev/infra/Dockerfile
Comment (Yikun) [inline dev/infra/Dockerfile:1]: cleanup(rebase) this when #37003 ready

File: .github/workflows/build_and_test.yml
Comment (Yikun) [inline .github/workflows/build_and_test.yml:867]: https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow

need to define secrets in main workflow

**Comment by Yikun** [comment]
- recompute is not blocked by the recovery *** FAILED *** (1 minute, 6 seconds)

Failure due to unrelated error, I will do a recheck after https://github.com/apache/spark/pull/37103 merged.

**Comment by Yikun** [comment]
Retriggered: https://github.com/Yikun/spark/runs/7212330062

File: .github/workflows/build_and_test.yml
Comment (Yikun) [inline .github/workflows/build_and_test.yml:274]: TODO: need a test for schedule job before ready to go

File: .github/workflows/build_and_test.yml
Comment (Yikun) [inline .github/workflows/build_and_test.yml:266]: TODO: need a test for schedule job before ready to go

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml]: Thanks for explanation.

**Review by HyukjinKwon — APPROVED** [review]

File: .github/workflows/build_and_test.yml
Comment (Yikun) [reply .github/workflows/build_and_test.yml:274]: `github.repository_owner`: `apache` in here according to [here](https://github.com/search?q=org%3Aapache+%22github.repository_owner%22&type=code)
`github.actor`: The latest user who create the schedule trigger will be the actor for the schedule workflow. `Yikun` in here

So, this may bring some issue on apace/spark job (such as schedule, push to master/branches). We can use the `image-cache` directly for jobs under apache/spark. 

[1] https://github.community/t/who-will-be-the-github-actor-when-a-workflow-runs-on-a-schedule/17369

File: .github/workflows/build_and_test.yml
Comment (Yikun) [inline .github/workflows/build_and_test.yml:266]:
```suggestion
          REPO_OWNER=$(echo "${{ github.actor }}" | tr '[:upper:]' '[:lower:]')
```

File: .github/workflows/build_and_test.yml
Comment (Yikun) [inline .github/workflows/build_and_test.yml:257]: Skip the schedule job in here.

File: .github/workflows/build_and_test.yml
Comment (Yikun) [inline .github/workflows/build_and_test.yml:316]: Use cache-image directly for schedule job.

File: .github/workflows/build_and_test.yml
Comment (Yikun) [reply .github/workflows/build_and_test.yml:274]: Case 1: The developer push a commit: ==> OK 
- github.actor: `yikun`
- generate and use image like `yikun/apache-spark-ci-image:master-xxx`

Case 2: The apache/spark merge a commit in master: ==> OK
- github.actor: `HyukjinKwon`
- generate and use image like `hyukjinkwon/apache-spark-ci-image:master-xxx`

Case 3: The scheduler jobs for master job:  ==> ???
- Use master image cache directly (`spark/apache-spark-github-action-image-cache:master`)

Case 4: The scheduler jobs for branches job  ==> OK
- Use branches image cache directly (`dongjoon/apache-spark-github-action-image:20220207`, future is `spark/apache-spark-github-action-image-cache:branches-*`)

Case 5: The developer push a commit in branches: ==> OK
Case 6: The apache/spark merge a commit in branches:
- use branches workflow, no change by this pr

**Comment by Yikun** [comment]
Does all spark committers has right to push apache ghcr package? According to [cache image](https://github.com/apache/spark/actions/workflows/build_infra_images_cache.yml), I can confirm @HyukjinKwon has rights, but not sure any different with other commiters.

**Comment by HyukjinKwon** [comment]
Let's get this in and see how it gose.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by Yikun** [comment]
> Let's get this in and see how it gose.

Sure, I will monitor in recent days. and update results here:
Case 1: The developer push a commit: 
Case2: Merge a commit in apache/spark master, work as expected: https://github.com/apache/spark/runs/7247609069?check_suite_focus=true#step:8:204
Case 3: Scheduler job (master) TB updated
Case 4: Scheduler job (branch-xxx)TB updated

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:278]: You could remove this comment

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:280]: and remove this too

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:290]: we could inline this

File: .github/workflows/build_and_test.yml
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:254]: BTW, let's add a `name` here to make it look pretty (e.g., Base image build)

**Comment by HyukjinKwon** [comment]
Feel free to make a followup or a separate PR with a separate JIRA 👍

File: .github/workflows/build_and_test.yml
Comment (Yikun) [reply .github/workflows/build_and_test.yml:283]: If a PR changes the dockerfile and merged, this will help user update the dockerfile...so, I think it's needed?

Comment (HyukjinKwon) [reply .github/workflows/build_and_test.yml:283]: oh okay you're correct!

File: .github/workflows/build_and_test.yml
Comment (Yikun) [reply .github/workflows/build_and_test.yml:280]: then keep this

File: .github/workflows/build_and_test.yml
Comment (Yikun) [reply .github/workflows/build_and_test.yml:278]: then keep this

**Pull request closed without merging**
